### PR TITLE
PG 11 compatibility refresh

### PR DESCRIPTION
**Pull request by mtuncer**
Reflects recent changes in PG11 codebase to citus.

### CODE REVIEW

**Comment by jasonmp85** [comment]
Can you elaborate on the modifying test? I see you pushed up one more commit and checks seem to have passed…

**Comment by mtuncer** [comment]
PG11 tests did not. Perhaps I should make PG11 non-optional until we decide to merge this.

**Comment by onderkalaci** [comment]
@mtuncer and @jasonmp85 (cc: @marcocitus), PostgreSQL 11 has a good number of improvements for the partitioning support such as `Create Index on Partitioned Tables`, `Allow Unique Index on Partitioned Tables` or `Allow FKs on Partitioned Tables`. They probably should just work out of the box, but, it'd be great if we briefly test them.

Do you guys think that is it a good PR to add some regression tests to cover those changes or do you prefer to handle that later?

I think we should at least open an issue to not lose track of that. What are your thoughts? Thanks!

**Comment by jasonmp85** [comment]
@mtuncer any updates, or should I just keep testing this a bit…? I think we could get this merged relatively quickly with one more back-and-forth.

**Comment by mtuncer** [comment]
I was on with modifying test. Did not work on your feedback yet.

eta monday

**Comment by jasonmp85** [comment]
> I was on with modifying test

Any discoveries? Should I look into it?

**Comment by mtuncer** [comment]
not really. it works on mac and ubuntu but not on travis.

could you look into it if you have time ?

**Comment by mtuncer** [comment]
@onderkalaci 
> Do you guys think that is it a good PR to add some regression tests to cover those changes or do you prefer to handle that later?

We aimed to make sure  citus builds and regression tests pass using PG11 with this PR.  I think anything above this calls for a different item.

Maintaining regression tests for different version is not very trivial, require manual work upon each change.

We have  3 versions of expected output for 3 versions of postgres for one regression test in this PR.

**Comment by jasonmp85** [comment]
> it works on mac and ubuntu but not on travis […] could you look into it if you have time ?

I gave it a look Friday but didn't get anywhere; I'll keep looking into it today.

File: src/test/regress/sql/multi_partitioning.sql [hunk 1]
```diff
@@ -261,6 +261,7 @@ SELECT master_modify_multiple_shards('UPDATE partitioning_test_2009 SET time = t
 
 -- test CREATE INDEX
 -- CREATE INDEX on partitioned table - this will error out
+-- on earlier versions of postgres.
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/test/regress/sql/multi_partitioning.sql]: ok

File: src/backend/distributed/executor/multi_utility.c [hunk 2]
```diff
@@ -1683,7 +1683,8 @@ static bool
 IsSupportedDistributedVacuumStmt(VacuumStmt *vacuumStmt)
 {
 	const char *stmtName = (vacuumStmt->options & VACOPT_VACUUM) ? "VACUUM" : "ANALYZE";
-	LOCKMODE lockMode = (vacuumStmt->options & VACOPT_FULL) ? AccessExclusiveLock : ShareUpdateExclusiveLock;
+	LOCKMODE lockMode = (vacuumStmt->options & VACOPT_FULL) ? AccessExclusiveLock :
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/backend/distributed/executor/multi_utility.c]: formatting change only. output of citus_indent

File: src/test/regress/expected/multi_join_order_additional.out [hunk 3]
```diff
@@ -29,23 +29,9 @@ CREATE TABLE lineitem_hash (
 	l_shipmode char(10) not null,
 	l_comment varchar(44) not null,
 	PRIMARY KEY(l_orderkey, l_linenumber) );
-<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/test/regress/expected/multi_join_order_additional.out]: done

File: src/backend/distributed/executor/multi_utility.c [hunk 4]
```diff
@@ -1612,26 +1620,37 @@ ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
 	Oid relationId = InvalidOid;
 	List *taskList = NIL;
 	bool supportedVacuumStmt = false;
+	RangeVar *relation = NULL;
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/backend/distributed/executor/multi_utility.c]: changed them accordingly.  Found and fixed a bug in the new code along the way

File: src/backend/distributed/executor/multi_utility.c [hunk 5]
```diff
@@ -1713,12 +1790,19 @@ VacuumTaskList(Oid relationId, VacuumStmt *vacuumStmt)
...
-	const char *columnNames = DeparseVacuumColumnNames(vacuumStmt->va_cols);
+	const char *columnNames = NULL;
 	const int vacuumPrefixLen = vacuumString->len;
 	Oid schemaId = get_rel_namespace(relationId);
 	char *schemaName = get_namespace_name(schemaId);
 	char *tableName = get_rel_name(relationId);
 
+#if (PG_VERSION_NUM >= 110000)
+	VacuumRelation *vacuumRelation = (VacuumRelation *) linitial(vacuumStmt->rels);
+	columnNames = DeparseVacuumColumnNames(vacuumRelation->va_cols);
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/backend/distributed/executor/multi_utility.c]: This code had a bug that it expects a single relation in the vacuum statement. fixed and applied your recommendation.

File: src/test/regress/expected/multi_multiuser.out [hunk 6]
```diff
@@ -110,15 +118,15 @@ HINT:  Run the command with a superuser.
...
+ERROR:  permission denied for table test
 EXECUTE prepare_select;
  count 
 -------
      2
 (1 row)
 
 INSERT INTO test VALUES (2);
-ERROR:  permission denied for relation test
+ERROR:  permission denied for table test
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/test/regress/expected/multi_multiuser.out]: I think we are stuck with it. Hopefully, it will be fine in few citus major releases.

File: src/backend/distributed/utils/statistics_collection.c [hunk 7]
```diff
@@ -249,7 +249,11 @@ CheckForUpdatesCallback(char *contents, size_t size, size_t count, void *userDat
 	{
 		Datum responseCStringDatum = CStringGetDatum(responseNullTerminated->data);
 		Datum responseJasonbDatum = DirectFunctionCall1(jsonb_in, responseCStringDatum);
+#if (PG_VERSION_NUM >= 110000)
+		responseJsonb = DatumGetJsonbP(responseJasonbDatum);
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/backend/distributed/utils/statistics_collection.c]: ok

**Comment by mtuncer** [comment]
@jasonmp85 I completed working on your feedback. I was also able to get the crash in with_modifying test. It was an assert.  I temporarily disabled that test again.
Will work on resolving that one next.

File: src/backend/distributed/executor/multi_utility.c [hunk 8]
```diff
@@ -1661,23 +1687,55 @@ ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
  * statement needs distributed execution but contains unsupported options.
  */
 static bool
-IsSupportedDistributedVacuumStmt(Oid relationId, VacuumStmt *vacuumStmt)
+IsSupportedDistributedVacuumStmt(VacuumStmt *vacuumStmt, List *vacuumRelationIdList)
 {
 	const char *stmtName = (vacuumStmt->options & VACOPT_VACUUM) ? "VACUUM" : "ANALYZE";
 	bool distributeStmt = false;
+	ListCell *relationIdCell = NULL;
+	int distributedRelationCount = 0;
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: TBH I'd just assign the `list_length` to `distributedRelationCount` right here and check for zero in the first `if` statement (with a suitable comment explaining `0` means _all relations_.

File: src/backend/distributed/executor/multi_utility.c [hunk 9]
```diff
@@ -1661,23 +1687,55 @@ ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
...
+		Oid relationId = lfirst_oid(relationIdCell);
+		if (OidIsValid(relationId) && IsDistributedTable(relationId))
+		{
+			distributedRelationCount++;
+		}
 	}
-	else if (!EnableDDLPropagation)
+
+	/*
+	 * There can only be one distributed relation in the vacuum statement.
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: I'd rephrase this as "we currently only support a single distributed relation in the `VACUUM` statement, because there _can_ be more than one, we just don't support it.

File: src/backend/distributed/executor/multi_utility.c [hunk 10]
```diff
@@ -1661,23 +1687,55 @@ ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
...
+	 * along with other relations.
+	 */
+	if (distributedRelationCount == 0)
+	{
+		/* Nothing to do here; there is no distributed relation */
+		return false;
+	}
+	else if (distributedRelationCount > 0 && vacuumedRelationCount > 1)
+	{
+		ereport(ERROR, (errmsg("vacuuming multiple tables at the same time "
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: I think I remembered why everything should be `WARNING`… We've already run the `VACUUM` against the local table at this point, right? But `VACUUM` is _never_ transactional, so throwing an error doesn't actually roll back anything, so I used a `WARNING`. For consistency, can we keep that behavior?

Comment (jasonmp85) [reply src/backend/distributed/executor/multi_utility.c]: Also `VACUUM`ing multiple tables at the same time _is_ supported, just not by Citus. Shouldn't this be phrased like "`VACUUM`ing a distributed table at the same time as other tables is not supported" or something?

File: src/include/distributed/version_compat.h [hunk 11]
```diff
@@ -44,6 +47,102 @@ PathNameOpenFilePerm(FileName fileName, int fileFlags, int fileMode)
...
+	{
+		VacuumRelation *vacuumRelation = (VacuumRelation *) lfirst(vacuumRelationCell);
+		vacuumList = lappend(vacuumList, vacuumRelation->relation);
+	}
+
+	return vacuumList;
+}
+
+
+static inline List *
```
Comment (jasonmp85) [inline src/include/distributed/version_compat.h]: Needs a comment.

File: src/include/distributed/version_compat.h [hunk 12]
```diff
@@ -44,6 +47,102 @@ PathNameOpenFilePerm(FileName fileName, int fileFlags, int fileMode)
...
+VacuumColumnList(VacuumStmt *vacuumStmt, int relationIndex)
+{
+	VacuumRelation *vacuumRelation = NULL;
+
+	Assert(relationIndex >= 0 && list_length(vacuumStmt->rels) > relationIndex);
```
Comment (jasonmp85) [inline src/include/distributed/version_compat.h]: I could be wrong but I thought many of the list utilities already have asserts similar to this?

File: .travis.yml [hunk 13]
```diff
@@ -19,10 +19,8 @@ matrix:
     - env: PGVERSION=9.6
     - env: PGVERSION=10
     - env: PGVERSION=11
-  allow_failures:
-    - env: PGVERSION=11
 before_install:
-  - git clone -b v0.7.6 --depth 1 https://github.com/citusdata/tools.git
+  - git clone -b pg11-patch --depth 1 https://github.com/citusdata/tools.git
```
Comment (jasonmp85) [inline .travis.yml]: Did you open a pull request for the tools change? I can merge it (I need to fix the Homebrew error anyhow).

**Comment by jasonmp85** [comment]
If you can add in whatever remaining feedback you want to address and figure out the test and I can probably merge this in tomorrow after a final look.

File: .travis.yml [hunk 14]
```diff
@@ -22,7 +22,7 @@ matrix:
   allow_failures:
     - env: PGVERSION=11
 before_install:
-  - git clone -b v0.7.6 --depth 1 https://github.com/citusdata/tools.git
+  - git clone -b pg11-patch --depth 1 https://github.com/citusdata/tools.git
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply .travis.yml]: Debian patch seems to be fixed. It can be applied now, Reverting this change.

File: src/test/regress/sql/multi_join_order_additional.sql [hunk 15]
```diff
@@ -10,9 +10,9 @@ SET citus.next_shard_id TO 650000;
 SET citus.explain_distributed_queries TO off;
 SET citus.log_multi_join_order TO TRUE;
 SET citus.task_executor_type = 'task-tracker'; -- can't explain all queries otherwise
-SET client_min_messages TO DEBUG2;
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/test/regress/sql/multi_join_order_additional.sql]: This was due to different output in PG11. Since it was not directly related to test, I decided to hide it.

File: src/backend/distributed/commands/create_distributed_table.c [hunk 16]
```diff
@@ -1103,8 +1103,14 @@ CreateTruncateTrigger(Oid relationId)
 	trigger->whenClause = NULL;
 	trigger->isconstraint = false;
 
+#if (PG_VERSION_NUM >= 110000)
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/backend/distributed/commands/create_distributed_table.c]: I did not like the fix. Having a function with the same name is not an option (breaks other places). Had to introduce a macro with a new name.

File: src/backend/distributed/executor/multi_utility.c (hunk 8 above)
Comment (mtuncer) [reply src/backend/distributed/executor/multi_utility.c]: ok

File: src/backend/distributed/utils/maintenanced.c [hunk 17]
```diff
@@ -276,7 +276,11 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		 databaseOid, myDbData->userOid);
 
 	/* connect to database, after that we can actually access catalogs */
+#if (PG_VERSION_NUM >= 110000)
+	BackgroundWorkerInitializeConnectionByOid(databaseOid, myDbData->userOid, 0);
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/backend/distributed/utils/maintenanced.c]: added a``` #define``` in ```version_compat.h````

File: src/include/distributed/version_compat.h (hunk 11 above)
Comment (mtuncer) [reply src/include/distributed/version_compat.h]: added

File: src/test/regress/expected/isolation_partitioned_copy_vs_all.out [hunk 18]
```diff
@@ -231,7 +231,7 @@ step s1-copy: COPY partitioned_copy FROM PROGRAM 'echo 5, f, 5 && echo 6, g, 6 &
 step s2-table-size: SELECT citus_total_relation_size('partitioned_copy');
 citus_total_relation_size
 
-32768          
+0
```
[the remark this answers is not in the document]
Comment (mtuncer) [reply src/test/regress/expected/isolation_partitioned_copy_vs_all.out]: just how PG11 behaves on partitioned tables

**Comment by mtuncer** [comment]
@jasonmp85 There are 2 outstanding issues left
1 - with_modifying test is disabled. It crashed with and assert when run in assert enabled build
2 - there is a build failure on travis (PR build). This is due to use of some recently added api.

**Comment by lithp** [comment]
Also, Windows was happy until 4efdeaa95b2, which appears to have [broken the build](https://ci.appveyor.com/project/lithp/citus/build/1.0.779-pg11_compat)

**Comment by jasonmp85** [comment]
@mtuncer I was able to fix the modifying test and also fixed a problem with the `_ext` method (it's been added to PostgreSQL but is not yet in any packaged releases). I like `compat` better as a suffix anyhow.

I'm heading to bed but if the other tests pass this is a :shipit:.

**Comment by mtuncer** [comment]
@jasonmp85 allowed PG11 build to fail in travis again.
merging it

File: src/backend/distributed/planner/distributed_planner.c [hunk 19]
```diff
@@ -1464,14 +1467,22 @@ HasUnresolvedExternParamsWalker(Node *expression, ParamListInfo boundParams)
 		if (boundParams && paramId > 0 && paramId <= boundParams->numParams)
 		{
 			ParamExternData *externParam = &boundParams->params[paramId - 1];
+			Oid paramType = externParam->ptype;
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c]: @mtuncer @jasonmp85 

This line causes an invalid memory read error on Valgrind because when `paramFetch ` is supplied, length of the `params []` could be zero, as documented in the Postgres code.

So, we shouldn't be accessing `externParam ` at this point.

It looks like a very similar line was already in Citus long before PG 11, see [commit](https://github.com/citusdata/citus/commit/6939cb8c561f896ece0a28b48cae7d8cb9f26ed3).  

Do you remember whether this is related with PG 11 or not? (e.g., was it always true that when `paramFetch` exists, the array size could be zero.)

Here is the valgrind error:
```
==31715== VALGRINDERROR-BEGIN
==31715== Invalid read of size 4
==31715==    at 0x6040559: HasUnresolvedExternParamsWalker (distributed_planner.c:1567)
==31715==    by 0x3E2A49: expression_tree_walker (nodeFuncs.c:2087)
==31715==    by 0x604050F: HasUnresolvedExternParamsWalker (distributed_planner.c:1601)
==31715==    by 0x3E2ACB: expression_tree_walker (nodeFuncs.c:2119)
==31715==    by 0x604050F: HasUnresolvedExternParamsWalker (distributed_planner.c:1601)
==31715==    by 0x3E2F4C: query_tree_walker (nodeFuncs.c:2263)
==31715==    by 0x6040599: HasUnresolvedExternParamsWalker (distributed_planner.c:1594)
==31715==    by 0x6041660: CreateDistributedPlannedStmt (distributed_planner.c:490)
==31715==    by 0x6041988: distributed_planner (distributed_planner.c:158)
==31715==    by 0x4404D3: planner (planner.c:260)
==31715==    by 0x50032E: pg_plan_query (postgres.c:832)
==31715==    by 0x50046D: pg_plan_queries (postgres.c:898)
```

Comment (onderkalaci) [reply src/backend/distributed/planner/distributed_planner.c]: Ok, I found the Postgres commit that changed the behaviour: https://github.com/postgres/postgres/commit/6719b238e8f0ea83c39d2b1728c7670cdaa34e06

I opened a PR (#2457) to reflect the change and make sure that we don't have invalid memory access.
